### ISSUE TITLE
Selective copy of white balance doesn't preserve "as shot"

### ISSUE THREAD

**Issue by fidergo-stephane-gourichon** [darktable-org/darktable#8365]
**Describe the bug**

"Selective copy" of the "white balance" module from a photo where it is set to "as shot" results, where pasted, to be set "user modified", not "as shot".

The impact is wrong white balance in photos, workaround is to manually walk **all** photos and click "as shot".

**To Reproduce**

* Open darktable, in preferences->processing->"auto-apply chromatic adaptation defaults" set to "modern", a.k.a. in darktablerc plugins/darkroom/chromatic-adaptation=modern. Exit darktable and open it again.
* (1) import dir with at least 2 RAWs (here NEF), 
* (2) open first, disable "color calibration", in "white balance" select "as shot", 
* (3) back to lighttable mode -> history stack -> "selective copy", "select none", select "white balance (enabled)" and "color calibration (disabled)", "ok", 
* (4) select all pictures, history stack -> "paste" 
* (5) open a picture different from the edited one. 

**Expected behavior**

"white balance" module shows: "settings: as shot" and all images show colors as set by the camera (as per the traditional white balance workflow of darktable ~3.2 and before)

Observed: "white balance" module shows: "settings: user modified", and photographs where the camera chose different white balance settings show wrong colors.

It can be argued that the copy-paste is kind of manually reverting to old color workflow. Still, copy-paste "as shot" from one image to any other (whether the camera has selected exactly the same white balance parameters or any other), is expected to produce "as shot" in destination images.

**Screenshots**

Not needed apparently. Please ask for if needed.

**Platform (please complete the following information):**
 - Darktable Version: 3.5.0 branch, exactly "3a1221288 view: finally using rowid to keep insertion order."
 - OS: Ubuntu 20.04
 - no OpenCL "0.039829 [opencl_init] FINALLY: opencl is NOT AVAILABLE on this system."
 - Intel Corporation UHD Graphics 8086:9b41

**Additional context**

 - Can you reproduce with another Darktable version? Not observed with 3.2 and the old-style color workflow (only white balance module, no "color calibration")
 - Can you reproduce with a RAW or Jpeg or both? Not applicable to JPEG, only RAW.
 - Are the steps above reproduce with a fresh edit (removing history)? Yes, fresh edit.
 - Attach an XMP if this is necessary -> as discussed on IRC, not needed
 - Did you compile Darktable yourself? If so which compiler was used, with what options? Compiled with 9.3.0-17ubuntu1~20.04
 - Is the issue still present using an empty/new config-dir? The issue does not occur on a new config directory. It is enough to activate in preferences->processing->"auto-apply chromatic adaptation defaults"->"modern", a.k.a. in darktablerc plugins/darkroom/chromatic-adaptation=modern to trigger the bug.

**Comment by fidergo-stephane-gourichon** [comment]
In my main profile, setting preferences->processing->"auto-apply chromatic adaptation defaults" to "legacy", prevents bug to appear with new imported photos, but does not actually fix the bug.

"cloudy" can be copy-pasted to all images and recognized as "cloudy" on all images. But copying "as shot" to all images becomes "user modified" on all images, and not the "as shot *of that particular image*".

Camera is Nikon D5500.

**Comment by johnny-bit** [comment]
Copying white balance means copying wb coefficients, not presets. "as shot" means "particular set of coeffs calculated by camera at point of shot" so if you copy "as shot" of one image that is NOT and will likely never be same as "as shot" of another image. There is no point in doing that.

**Comment by fidergo-stephane-gourichon** [comment]
Thanks @johnny-bit for your answer, yet don't be too fast in your conclusions. There might be no point in doing what you explain (which looks implementation-oriented and presumes the issue reporter does not understand), yet that was probably not what this is all about. It's about a actual use case that worked in previous versions of darktable and no longer work in recent versions. 

Several users (on IRC #darktable) got it. At least two, including me, experienced a situation where a number of photographs (possibly hundreds or more) had some processing applied, including specific crop, different denoising parameters, specific perspective parameters and whatever other modules. It happened that the white balance module was applied, but the settings were not satisfactory (a number of scenarios can lead to such a situation, e.g. pasting a wrong settings, or migrating from legacy color workflow to new color workflow). Pasting "as shot" from an image to other images is expected to set the white balance to *whatever the camera chose for each photo*. It might not fit the current implementation you have in mind, still it's a pretty relevant use case from a user point of view, so "no point in doing that" misses the point.

With such a reset-copy-paste workflow, such problems are neatly solved. Without it, the user has no better choice than editing every image one by one and press the module reset button. I've been using darktable for about 8 years and loved it because such repetitive chore was *never* needed. Current (master branch) state is a major regression from this use case point of view.

I had a discussion on IRC the same evening with a few DT developers, that clarified a lot. To put it short, the old code could do that, but it was more an accidental code behavior than an actually elaborated feature. The fact that the "auto white balance from camera" setting in the old white balance module is the "default behavior" when module has no parameter for this photograph, and for some reason that could be copy-pasted. There was much internal refactoring for future-proofing and DT does not currently has a clean, actually intended, generic way to "paste a module reset order" from one photo to other(s). The conclusion from a major DT developer was that it may be interesting to implement. 

Examples where it is applicable:

* reset "crop and rotate" to whatever different dimensions and orientation *each* photograph is
* reset "perspective"
* reset "lens correction" to apply using whatever *different* processing (distorsion, chromatic aberration, etc) is actually best for each photograph
* reset "profiled denoise" to whatever *different* sensor and iso *each* photograph was shot.
* reset "exposure", obeying whatever *different per-photograph bias* the camera may have set
* more are probably applicable

For each of the modules above, copy-pasting a "reset this module" order is a straightforward way to express a user intent, and according to the IRC conversation, it can be done in a generic way for all modules.

Remember also that, worse than the "particular set of coeffs as shot will not match another photo", be aware that collections of photographs may be from different cameras (and often are in my case). Yet this is *not a problem* with the proposed workflow. Whenever darktable is able to open a photograph and apply default processing for a module, resetting the same module will do something relevant for that photograph. I wish that feature of darktable that I used so often in the last 8 years evolves from an implementation accident into a supported feature. I read on IRC that some modules (e.g. lens correction) had special code to specifically to support such a workflow. If a generic solution works for all modules, it might even simplify the code base ?

**Comment by fidergo-stephane-gourichon** [comment]
Might be related to #8861.

**Comment by fidergo-stephane-gourichon** [comment]
Hi,

TL;DR: copy-pasting consistent white balance settings from one raw to another used to produce inconsistent results, now produces expected consistent results. Bug actually fixed. Thanks.

Tested again the behavior on commit c4789e899e a.k.a. `release-4.3.0-1033-gc4789e899e`.

> **Reproduce**

> Open darktable, in preferences->processing->"auto-apply chromatic adaptation defaults" set to "modern", a.k.a. in darktablerc plugins/darkroom/chromatic-adaptation=modern. Exit darktable and open it again.

In preference section *preferences->processing*, the item *auto-apply chromatic adaptation defaults* value *modern* (in `darktablerc` `plugins/darkroom/chromatic-adaptation=modern`) looks like it has become *auto-apply pixel workflow defaults* value *scene-referred (filmic)* (in `darktablerc` `plugins/darkroom/workflow=scene-referred (filmic)`).

In both cases the default value was the value that is suitable to test for the bug.

> **Expected behavior**

> "white balance" module shows: "settings: as shot" and all images show colors as set by the camera (as per the traditional white balance workflow of darktable ~3.2 and before)

> Observed: "white balance" module shows: "settings: user modified", and photographs where the camera chose different white balance settings show wrong colors.

I now get the "expected behavior".


I did not get the expected behavior in a similar test on the lens correction module.  Since that is outside the scope of this bug, I let some time to experiment a little more before, perhaps, discussing that again. I noticed, when copying, a "reset" checkbox which looks like it does most of what I expected and should cover most cases (although I did not see it mentioned on https://docs.darktable.org/usermanual/4.0/en/module-reference/utility-modules/lighttable/history-stack/ ).
